chronos/asyncsync.nim

```diff
@@ -610,3 +626,162 @@ template emitWait*[T](bus: AsyncEventBus, event: string,
   ## wait until all the subscribers/waiters will receive notification about
   ## event.
   emitWait(bus, event, data, getSrcLocation())
+
+proc `==`(a, b: EventQueueKey): bool {.borrow.}
+
+proc compact[T](ab: AsyncEventQueue[T]) =
+  if len(ab.readers) > 0:
+    let minOffset = ab.readers[0].offset
+    doAssert(minOffset >= ab.offset)
+    if minOffset > ab.offset:
+      let delta = minOffset - ab.offset
+      ab.queue.shrink(fromFirst = delta)
+      ab.offset += delta
+  else:
+    ab.queue.clear()
+
+proc getReaderIndex[T](ab: AsyncEventQueue[T], key: EventQueueKey): int =
+  for index, value in ab.readers.pairs():
+    if value.key == key:
+      return index
+  -1
+
+proc newAsyncEventQueue*[T](limitSize = 0): AsyncEventQueue[T] =
+  ## Creates new ``AsyncEventBus`` maximum size of ``limitSize`` (default is
+  ## ``0`` which means that there no limits).
+  ##
+  ## When number of events emitted exceeds ``limitSize`` - emit() procedure
+  ## will discard new events, consumers which has number of pending events
+  ## more than ``limitSize`` will get ``AsyncEventQueueFullError``
+  ## error.
+  doAssert(limitSize >= 0, "Limit size should be non-negative integer")
+  let queue =
+    if limitSize == 0:
+      initDeque[T]()
+    elif isPowerOfTwo(limitSize + 1):
+      initDeque[T](limitSize + 1)
+    else:
+      initDeque[T](nextPowerOfTwo(limitSize + 1))
+  AsyncEventQueue[T](counter: 0'u64, queue: queue, limit: limitSize)
+
+proc len*[T](ab: AsyncEventQueue[T]): int =
+  len(ab.queue)
+
+proc register*[T](ab: AsyncEventQueue[T]): EventQueueKey =
+  inc(ab.counter)
+  let reader = EventQueueReader(key: EventQueueKey(ab.counter),
+                                offset: ab.offset + len(ab.queue))
+  ab.readers.add(reader)
+  EventQueueKey(ab.counter)
+
+proc unregister*[T](ab: AsyncEventQueue[T], key: EventQueueKey) =
+  let index = ab.getReaderIndex(key)
+  if index >= 0:
+    let reader = ab.readers[index]
+    # Completing pending Future to avoid deadlock.
+    if not(isNil(reader.waiter)) and not(reader.waiter.finished()):
+      reader.waiter.complete()
+    ab.readers.delete(index)
+    ab.compact()
+
+proc emit*[T](ab: AsyncEventQueue[T], data: T) =
+  if len(ab.readers) > 0:
+    # We enqueue `data` only if there active reader present.
+    let couldEmit =
+      if ab.limit == 0:
+        true
+      else:
+        # Because ab.readers is sequence sorted by `offset`, we will apply our
+        # limit to the most recent consumer.
+        if ab.limit + (ab.readers[^1].offset - ab.offset) < len(ab.queue):
+          false
+        else:
+          true
+
+    if couldEmit:
+      ab.queue.addLast(data)
```

Comment:
ok, so here, for bounded queues, I think we need to clean up readers that are out-of-bounds / slow / stopped reading and call `compact`, or the queue will grow as if it were unbounded - consider:

```
let eventQueue = newAsyncEventQueue[seq[int]]()
let
  fast = eventQueue.register()
  slow = eventQueue.register()
  slowF = eventQueue.waitEvents(slow)
for i in 0..<1000:
  let fastF = eventQueue.waitEvents(fast)
  eventQueue.emit(newSeq[int](100*1024*1024))
  discard waitFor fastF
  GC_fullCollect()

  debugEcho GC_getStatistics()
```

this program should never keep more than one entry in its data structure, but keeps allocating more and more memory because `waitEvents` will not run until the slow reader schedules itself (consider for example if it's stuck in a socket `write` which takes minutes to timeout)

Reply:
Your example do not use bounded queue, is it intentional?

Reply:
oops, old version of example - edited to adds bounds, memory usage still explodes